core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -231,6 +225,16 @@ private Object handleTxCommand(TxInvocationContext ctx, TransactionBoundaryComma
          }
       } else {
          if (currentTopologyId() > command.getTopologyId()) {
+            //in some cases, the node will not retry the transaction in this node and then it leaks.
+            //if it is safe, we rollback the transaction. If the coordinator retries on this node, no side effect is created.
```

Comment:
I think it's a good idea to keep the prepare command idempotent. IIRC if there is a merge, JGroups may retransmit old broadcasted messages to the other partition, long after the original sync broadcast ended. So you might receive a prepared tx via state transfer, then you might receive a prepare tx with an old topology id that rolls back the tx and leaves you with nothing to commit.